### PR TITLE
osd/OSD: fix trim_maps() - possible leak on `skip_maps`

### PR DESCRIPTION
**Pull request by Matan-B**
~Blocked by: #52339~

Fixes: https://tracker.ceph.com/issues/61962

Example map gap scenario: https://gist.github.com/Matan-B/9b0eed8daee3bd6c3216bd3b6d11e8fb

Note: In a separate PR the leak can be addressed using an asock command to re-trim leaked osdmaps if those exist in the cluster - #53227

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/osd_types.cc [hunk 1]
```diff
@@ -5721,21 +5721,25 @@ void OSDSuperblock::encode(ceph::buffer::list &bl) const
...
   if (struct_v < 3) {
     string magic;
     decode(magic, bl);
   }
   decode(cluster_fsid, bl);
   decode(whoami, bl);
   decode(current_epoch, bl);
-  decode(oldest_map, bl);
-  decode(newest_map, bl);
+  if (struct_v >= 11) {
```
Comment (athanatos) [inline src/osd/osd_types.cc:5738]: This looks wrong -- these fields exist regardless of version.  Have you confirmed with the ceph-deencoder/object corpus that this works correctly with previous version superblocks?

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -5765,6 +5769,9 @@ void OSDSuperblock::decode(ceph::buffer::list::const_iterator &bl)
   } else {
     cluster_osdmap_trim_lower_bound = 0;
   }
+  if (struct_v >= 11) {
+    decode(maps, bl);
+  }
```
Comment (athanatos) [inline src/osd/osd_types.cc:5774]:
```
} else {
  maps.insert(oldest_map, newest_map - oldest_map);  // or w/e the correct interval_set usage is
}
```

File: src/osd/osd_types.cc (hunk 1 above)
Comment (athanatos) [reply src/osd/osd_types.cc:5738]: Should be unconditionally

```
    epoch_t oldest_map, newest_map;
    decode(oldest_map, bl);
    decode(newest_map, bl);
```

You need them below to initialize maps if the encoding version < 11.  Ensure that the ceph-object-corpus/ceph-deencoder tests catch this mistake before you fix it.

File: src/os/Transaction.h [hunk 3]
```diff
@@ -291,12 +291,6 @@ class Transaction {
   Transaction(const Transaction& other) = default;
   Transaction& operator=(const Transaction& other) = default;
 
-  Transaction claim_and_reset() {
-    auto ret = Transaction();
-    std::swap(*this, ret);
-    return ret;
-  }
```
Comment (athanatos) [inline src/os/Transaction.h:298]: I'd actually like to retain this -- there are situations where it would be useful.

**Review by athanatos — CHANGES_REQUESTED** [review]
Other than the encode/decode compat issue, I think this is pretty elegant.

There is another approach.  Instead of recording exactly which maps are on disk, we could retain OSDSuperblock::newest_map and OSDSuperblock::cluster_osdmap_trim_lower_bound.  The OSD would always have all maps in the set [cluster_osdmap_trim_lower_bound, newest_map] but also some number of not necessarily contiguous untrimmed maps < cluster_osdmap_trim_lower_bound.  Instead of using state in the superblock (oldest_map or maps), OSD::trim_maps() would simply list the first N maps in the meta collection (assuming they sort correctly -- check the object names to confirm) and remove them up to the min(service.map_cache.cached_key_lower_bound(), cluster_osdmap_trim_lower_bound).  The disadvantage here would be needing to do a collection list in OSD::trim_maps(), but the advantage would be that it would naturally clean up any previously leaked maps and would remove the ability for the superblock to disagree with stored maps (at least in this way).

**Comment by Matan-B** [comment]
> Other than the encode/decode compat issue, I think this is pretty elegant.
> 
> There is another approach. Instead of recording exactly which maps are on disk, we could retain OSDSuperblock::newest_map and OSDSuperblock::cluster_osdmap_trim_lower_bound. The OSD would always have all maps in the set [cluster_osdmap_trim_lower_bound, newest_map] but also some number of not necessarily contiguous untrimmed maps < cluster_osdmap_trim_lower_bound. Instead of using state in the superblock (oldest_map or maps), OSD::trim_maps() would simply list the first N maps in the meta collection (assuming they sort correctly -- check the object names to confirm) and remove them up to the min(service.map_cache.cached_key_lower_bound(), cluster_osdmap_trim_lower_bound). The disadvantage here would be needing to do a collection list in OSD::trim_maps(), but the advantage would be that it would naturally clean up any previously leaked maps and would remove the ability for the superblock to disagree with stored maps (at least in this way).

I like the idea of cleaning up any previously leaked maps in this approach. Moreover, oldest_map could also be retained merely for future debugging (mostly, where `map_cache.cached_key_lower_bound()` is at).
However, it seems that osdmaps aren't sorted in the meta collections:
```
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff0%f0%e8%26%21osdmap.13%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff1%80%e8%26%21osdmap.20%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff2%e0%e8%26%21osdmap.24%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff3%40%e8%26%21osdmap.17%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff4%c0%e8%26%21osdmap.19%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff5p%e8%26%21osdmap.11%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff5%a0%e8%26%21osdmap.26%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff60%e8%26%21osdmap.15%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff7%60%e8%26%21osdmap.22%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff8%20%e8%26%21osdmap.25%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff8%90%e8%26%21osdmap.10%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
O       %7f%7f%ff%ff%ff%ff%ff%ff%ff9%c0%e8%26%21osdmap.18%21%3d%00%00%00%00%00%00%00%00%ff%ff%ff%ff%ff%ff%ff%ffo
```

Same when listing the objects using `collection_list`:
```
[#-1:3058e826:::osdmap.28:0#,#-1:30f0e826:::osdmap.13:0#,#-1:3138e826:::osdmap.35:0#,
#-1:3180e826:::osdmap.20:0#,#-1:3248e826:::osdmap.39:0#,#-1:32e0e826:::osdmap.24:0#,
#-1:3340e826:::osdmap.17:0#,#-1:3398e826:::osdmap.31:0#,#-1:3478e826:::osdmap.33:0#,
#-1:34c0e826:::osdmap.19:0#,#-1:3508e826:::osdmap.40:0#,#-1:3570e826:::osdmap.11:0#,
#-1:35a0e826:::osdmap.26:0#,#-1:3630e826:::osdmap.15:0#,#-1:3760e826:::osdmap.22:0#,
#-1:37b8e826:::osdmap.37:0#,#-1:3820e826:::osdmap.25:0#,#-1:3890e826:::osdmap.10:0#,
#-1:3978e826:::osdmap.32:0#,#-1:39c0e826:::osdmap.18:0#,#-1:3a38e826:::osdmap.36:0#,
#-1:3a80e826:::osdmap.21:0#,#-1:3b30e826:::osdmap.14:0#,#-1:3bd8e826:::osdmap.29:0#,
#-1:3c18e826:::osdmap.30:0#,#-1:3cb0e826:::osdmap.16:0#,#-1:3d48e826:::osdmap.38:0#,
#-1:3de0e826:::osdmap.23:0#,#-1:3e08e826:::osdmap.41:0#,#-1:3e70e826:::osdmap.12:0#,
#-1:3ea0e826:::osdmap.27:0#,#-1:3ff8e826:::osdmap.34:0# ...
```

The current approach with the cleanup as an additional tool is preferable in this case, right?

**Comment by athanatos** [comment]
Ah:

```
  static ghobject_t get_osdmap_pobject_name(epoch_t epoch) {
    char foo[20];
    snprintf(foo, sizeof(foo), "osdmap.%d", epoch);
    return ghobject_t(hobject_t(sobject_t(object_t(foo), 0)));
  }
```

```
 /* Do not use when a particular hash function is needed */
  explicit hobject_t(const sobject_t &o) :
    oid(o.oid), snap(o.snap), max(false), pool(POOL_META) {
    set_hash(std::hash<sobject_t>()(o));
  }
```

we use the hobject_t(const sobject_t &) constructor to create OSDMap object names, so we get a random hash which doesn't sort usefully.  Unfortunate.  The approach in this PR will work.

File: src/osd/osd_types.cc (hunk 1 above)
Comment (Matan-B) [reply src/osd/osd_types.cc:5738]: Verified with `test/encoding/readable.sh`:
Before suggested change:
```
$ ../src/test/encoding/readable.sh > read_res.tst

$ tail -n 1 read_res.tst                              
FAILED 96 / 9438 tests.

$ cat read_res.tst | grep failed | grep Super |  wc -l
96
```
```

```
---

After change:
```
$ ../src/test/encoding/readable.sh > read_res2.tst

$ tail -n 1 read_res.tst                              
passed 9534 tests.
```

**Comment by Matan-B** [comment]
@athanatos, changeset: Applied suggested changes and rebased.

**Comment by Matan-B** [comment]
jenkins test make check

File: qa/suites/rados/thrash/thrashers/mapgap.yaml [hunk 4]
```diff
@@ -18,10 +18,12 @@ overrides:
         osd scrub max interval: 120
         osd scrub during recovery: false
         osd max backfills: 6
```
Comment (athanatos) [inline qa/suites/rados/thrash/thrashers/mapgap.yaml:20]: commit message prbabilty->probability

**Review by athanatos — APPROVED** [review]

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/REEF#httpstrellocomcT9bt0Ezq1871-wip-yuri4-testing-2023-10-31-1447-old-wip-yuri4-testing-2023-10-30-1117-old-wip-yuri4-testing-2023-10-27-1200-old-wip-yuri4-test

**Comment by rzarzynski** [comment]
Let's bake in `main` for awhile.

**Comment by jianwei1216** [comment]
https://github.com/ceph/ceph/pull/54686

**Pull request merged**